Updates to reflect service worker registration being opt-in.

jeffposnick opened the pull request:
R: @gaearon @Timer etc.

Fixes #3864

There are some changes to the docs needed if/when `navigateFallback` is re-enabled in the service worker configuration, but those are tracked in https://github.com/facebook/create-react-app/pull/3870 and I'm guessing that @piotr-cz will tackle them?

jeffposnick commented:
(Re: the CLA, it has been previously addressed outside of the normal submission process.)

Timer commented on packages/react-scripts/template/README.md:
```diff
@@ -74,8 +74,8 @@ You can find the most recent version of this guide [here](https://github.com/fac
 - [Publishing Components to npm](#publishing-components-to-npm)
-- [Making a Progressive Web App](#making-a-progressive-web-app)
-  - [Opting Out of Caching](#opting-out-of-caching)
+- [Making a Progressive Web App](#making-Making a Progressive Web Appa-progressive-web-app)
```
Oops

viankakrisna commented on packages/react-scripts/template/README.md:
```diff
@@ -1825,10 +1825,35 @@ Create React App doesn't provide any built-in functionality to publish a compone
+
+>Note: In the current version of `create-react-app`, opting-in is required for
+all developers who want offline-first behavior, even if you've previously
+deployed a version of your web app which used a service worker. If you redeploy
```
I don't think this is the right place to say redeploy because the opt-in only affect the people who generate a new project with CRA. The people that are upgrading to v2 from existing v1 generated project would still have the old registration code. (if they don't also update the file manually, it is)

jeffposnick replied:
Ah, so the local `src/index.js` is left alone when someone updates `create-react-app` from v1 to v2? I've got to be honest that I'm not familiar with the upgrade process 😄 

If that's the case, I'll update the text. Thanks!

iansu replied:
Yes, all files in `src` are left alone when upgrading. "Upgrading" is really just updating a dependency in `package.json`. The template code in `src` is only used when creating a new project.

viankakrisna replied:
usually, there would be a guide in the release like https://github.com/facebook/create-react-app/releases/tag/v1.0.8 to migrate the files in src, which I think is the more appropriate place for it.

piotr-cz commented:
#3870 is just revert of #3419

piotr-cz commented:
@jeffposnick I think it'll be more appropriate if I'd close #3870 and you'd handle the situation in one PR (revert of #3419 and updating the docs)

piotr-cz commented:
@jeffposnick Thanks, I've closed #3870

jeffposnick commented:
I've moved @piotr-cz's changes from #3870 (reinstating `navigateFallback`) into this PR.

I've also added a config option to [disable `skipWaiting`](https://developers.google.com/web/fundamentals/primers/service-workers/lifecycle#skip_the_waiting_phase), as it leads to better compatibility with lazy-loading at the expense of potentially extended the lifetime of previously cached data (until you close the final window/tab that had an old version of your web app open). Since a developer who explicitly opts-in to service worker support is less likely to be confused about showing cached data, it seemed like a good compromise.

jliebrand commented:
Hey guys - any update on this PR?

raiprabh commented:
Hey everyone, any estimate when this will be merged?
